File: 0304-range-sum-query-2d-immutable/0304-range-sum-query-2d-immutable.py
class NumMatrix:

    def __init__(self, matrix: List[List[int]]):
        rows = len(matrix) + 1
        cols = len(matrix[0]) + 1
        self.pre = [[0]*cols for _ in range(rows)]
        for r in range(1, rows):
            for c in range(1, cols):
                self.pre[r][c] = self.pre[r-1][c] + self.pre[r][c-1] - self.pre[r-1][c-1] + matrix[r-1][c-1]
                
    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        r2 = row2 + 1
        c2 = col2 + 1
        return self.pre[r2][c2] + self.pre[row1][col1] - self.pre[row1][c2] - self.pre[r2][col1]
        # Your NumMatrix object will be instantiated and called as such:
        # obj = NumMatrix(matrix)
        # param_1 = obj.sumRegion(row1,col1,row2,col2)

File: 0304-range-sum-query-2d-immutable/0304-range-sum-query-2d-immutable.cpp
class NumMatrix {
public:
    vector <vector<int>> pre;
    NumMatrix(vector<vector<int>>& matrix) {
        vector <int> row_pre;
        pre.push_back(vector<int>(matrix[0].size() + 1));
        long sum_;
        for(auto r: matrix){
            row_pre.clear();
            row_pre.push_back(0);
            sum_ = 0;
            for(auto n: r){
                sum_ += n;
                row_pre.push_back(sum_);
            }
            pre.push_back(row_pre);
        }
        for(int c = 1; c <= matrix[0].size(); ++c){
            sum_ = 0;
            for(int r = 1; r <= matrix.size(); ++r){
                sum_ += pre[r][c];
                pre[r][c] = sum_;
            }
        }
    }
    
    int sumRegion(int row1, int col1, int row2, int col2) {
        return pre[row2 + 1][col2 + 1] + pre[row1][col1] - pre[row2 + 1][col1] - pre[row1][col2 + 1];
    }
};

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix* obj = new NumMatrix(matrix);
 * int param_1 = obj->sumRegion(row1,col1,row2,col2);
 */